src/spline.c

```diff
@@ -164,3 +164,26 @@ void twin_path_curve(twin_path_t *path,
                              _twin_matrix_x(&path->state.matrix, x3, y3),
                              _twin_matrix_y(&path->state.matrix, x3, y3));
 }
+
+void twin_path_quadratic_curve(twin_path_t *path,
+                               twin_fixed_t x1,
+                               twin_fixed_t y1,
+                               twin_fixed_t x2,
+                               twin_fixed_t y2)
+{
+    twin_spoint_t p0 = _twin_path_current_spoint(path);
+    /* Convert quadratic to cubic control point */
+    /* CP1 = P0 + 2/3 * (P1 - P0) */
+    twin_sfixed_t dx1 = twin_fixed_to_sfixed(x1) - p0.x;
+    twin_sfixed_t dy1 = twin_fixed_to_sfixed(y1) - p0.y;
+    twin_sfixed_t cx1 = p0.x + twin_sfixed_mul(dx1, twin_int_to_sfixed(2) / 3);
+    twin_sfixed_t cy1 = p0.y + twin_sfixed_mul(dy1, twin_int_to_sfixed(2) / 3);
+    /* CP2 = P2 + 2/3 * (P1 - P2) */
+    twin_sfixed_t x2s = twin_fixed_to_sfixed(x2);
+    twin_sfixed_t y2s = twin_fixed_to_sfixed(y2);
+    twin_sfixed_t dx2 = twin_fixed_to_sfixed(x1) - x2s;
+    twin_sfixed_t dy2 = twin_fixed_to_sfixed(y1) - y2s;
+    twin_sfixed_t cx2 = x2s + twin_sfixed_mul(dx2, twin_int_to_sfixed(2) / 3);
+    twin_sfixed_t cy2 = y2s + twin_sfixed_mul(dy2, twin_int_to_sfixed(2) / 3);
+    _twin_path_scurve(path, cx1, cy1, cx2, cy2, x2s, y2s);
```

Comment:
I’m curious why the function `twin_path_quadratic_curve()` does not use the function `_twin_matrix_x()` and `_twin_matrix_y()`. 
Another question: Is there any difference compared to directly calculating the values `x1`, `y1`, `x2`, and `y2`, which are of type `twin_fixed_t`, and then calling `twin_path_curve(path, x1, y1, x2, y2, x3, y3)`?

Reply:
Is it because operations with `twin_fixed_t` are too complex?

Reply:
```c
void twin_path_quadratic_curve(twin_path_t *path,
                               twin_fixed_t x1,
                               twin_fixed_t y1,
                               twin_fixed_t x2,
                               twin_fixed_t y2)
{
    twin_spoint_t p0 = _twin_path_current_spoint(path);
    twin_fixed_t x1_ratio = twin_fixed_div(
        twin_fixed_mul(x1, twin_int_to_fixed(2)), twin_int_to_fixed(3));
    twin_fixed_t y1_ratio = twin_fixed_div(
        twin_fixed_mul(y1, twin_int_to_fixed(2)), twin_int_to_fixed(3));
    twin_path_curve(
        path,
        twin_fixed_div(twin_sfixed_to_fixed(p0.x), twin_int_to_fixed(3)) +
            x1_ratio,
        twin_fixed_div(twin_sfixed_to_fixed(p0.y), twin_int_to_fixed(3)) +
            y1_ratio,
        x1_ratio + twin_fixed_div(x2, twin_int_to_fixed(3)),
        y1_ratio + twin_fixed_div(y2, twin_int_to_fixed(3)), x2, y2);
}
```

Reply:
> I’m curious why the function `twin_path_quadratic_curve()` does not use the function `_twin_matrix_x()` and `_twin_matrix_y()`.

Yeah. I miss it, Thanks for the reminder.

>Another question: Is there any difference compared to directly calculating the values x1, y1, x2, and y2, which are of type twin_fixed_t, and then calling twin_path_curve(path, x1, y1, x2, y2, x3, y3)?

It's because the `p0` we get from `_twin_path_current_spoint` is translated by the `path->state.matrix`(I would like to call it the **absolute coordinate**) while `(x1, y1)`, `(x2, y2)` are not translated (call it the **relative coordinate**) and we should not mixed these two system coordinate in calculation.

Since we could't get the relative coordinate of `p0`(it would require inverse matrix of `path->state.matrix` or additional variable to store it), I calculate the point with absolute coordinate system and call `_twin_path_scurve ` to produce the curve.

I think the revised code would be like this?
```c
void twin_path_quadratic_curve(twin_path_t *path,
                               twin_fixed_t x1,
                               twin_fixed_t y1,
                               twin_fixed_t x2,
                               twin_fixed_t y2)
{
    twin_spoint_t p0 = _twin_path_current_spoint(path);
    /* Convert quadratic to cubic control point */
    twin_sfixed_t x1s = _twin_matrix_x(&path->state.matrix, x1, y1);
    twin_sfixed_t y1s = _twin_matrix_y(&path->state.matrix, x1, y1);
    twin_sfixed_t x2s = _twin_matrix_x(&path->state.matrix, x2, y2);
    twin_sfixed_t y2s = _twin_matrix_y(&path->state.matrix, x2, y2);
    /* CP1 = P0 + 2/3 * (P1 - P0) */
    twin_sfixed_t dx1 = x1s - p0.x;
    twin_sfixed_t dy1 = y1s - p0.y;
    twin_sfixed_t cx1 = p0.x + twin_sfixed_mul(dx1, twin_double_to_sfixed(2.0/3.0));
    twin_sfixed_t cy1 = p0.y + twin_sfixed_mul(dy1, twin_double_to_sfixed(2.0/3.0));
    /* CP2 = P2 + 2/3 * (P1 - P2) */
    twin_sfixed_t dx2 = x1s - x2s;
    twin_sfixed_t dy2 = y1s - y2s;
    twin_sfixed_t cx2 = x2s + twin_sfixed_mul(dx2, twin_double_to_sfixed(2.0/3.0));
    twin_sfixed_t cy2 = y2s + twin_sfixed_mul(dy2, twin_double_to_sfixed(2.0/3.0));
    _twin_path_scurve(path, cx1, cy1, cx2, cy2, x2s, y2s);
}
```

Reply:
> Since we could't get the relative coordinate of p0(it would require inverse matrix of path->state.matrix or additional variable to store it), I calculate the point with absolute coordinate system and call _twin_path_scurve  to produce the curve.

However, why it need to get the relative coordinate of p0? My idea is calculating the cubic spline's P1 and P2 which is in the absolute coordinate system first and then using `twin_path_curve()` to get the relative coordinate of p1, p2 and p3.

The function of `twin_path_curve` is to change the point to relative coordinate and then change to `twin_sfixed_t`.

The function of `twin_path_quadratic_curve()` is to creat the cubic spline's P1 and P2 which is in the absolute coordinate system by P1 of quadratic spline and then use `twin_path_curve()` is to change the point to relative coordinate and then change to `twin_sfixed_t`.
![spline (1)](https://github.com/user-attachments/assets/9d086edf-5860-4f6d-a11f-5573b2273709)

Reply:
Sorry, I didn't get it well.

`twin_path_curve` convert the input point with **relative coordinate**  and convert it with  `_twin_matrix_x` and `_twin_matrix_y` to point with **absolute coordinate** and pass it to  `_twin_path_scurve` and I just design `twin_path_quadratic_curve` with same style( convert **relative coordinate** input point and pass it with **absolute coordinate** to   `_twin_path_scurve`)
Maybe you could simply write your idea with code again?

Reply:
```c
void twin_path_quadratic_curve(twin_path_t *path,
                               twin_fixed_t x1,
                               twin_fixed_t y1,
                               twin_fixed_t x2,
                               twin_fixed_t y2)
{
    twin_spoint_t p0 = _twin_path_current_spoint(path);
    /* Calculate (2/3) * P1 first */
    x1 -= twin_fixed_div(x1, twin_int_to_fixed(3));
    y1 -= twin_fixed_div(y1, twin_int_to_fixed(3));
    /* Reuse the function twin_path_curve() that already implements "converting
     * relative coordinate input point and passing it with absolute coordinate
     * to _twin_path_scurve()." */
    return twin_path_curve(
        path,
        x1 + twin_fixed_div(twin_sfixed_to_fixed(p0.x), twin_int_to_fixed(3)),
        y1 + twin_fixed_div(twin_sfixed_to_fixed(p0.y), twin_int_to_fixed(3)),
        x1 + twin_fixed_div(x2, twin_int_to_fixed(3)),
        y1 + twin_fixed_div(y2, twin_int_to_fixed(3)), x2, y2);
}
```

Reply:
> It's because the p0 we get from _twin_path_current_spoint is translated by the path->state.matrix(I would like to call it the absolute coordinate) while (x1, y1), (x2, y2) are not translated (call it the relative coordinate) and we should not mixed these two system coordinate in calculation.

Just like I have mentioned, you couldn't calculate `p0` ( already translated to absolute coordinate by transition matrix) with `x1`, `x2`, `y1`, `y2`

When I applied rotation (make the transition matrix is not identity) to the `path` and use your code, the quadratic curve would be like incorrect
<img width="403" alt="image" src="https://github.com/user-attachments/assets/54a455e2-77c4-460a-9a63-a3105fedb59f">

That's why I said `twin_sfixed_to_fixed(p0.x`) needs to be replaced with an operation involving the inverse matrix, making the calculation with x1 and x2 reasonable.

Reply:
OK